compiler/codegen/OMRCodeGenerator.cpp

```diff
@@ -363,6 +363,78 @@ OMR::CodeGenerator::generateCodeFromIL()
    return false;
    }
 
+void
+OMR::CodeGenerator::insertGotoIntoLastBlock(TR::Block *lastBlock)
+   {
+   // If the last tree in the last block is not a TR_goto, insert a goto tree
+   // at the end of the block.
+   // If there is a following block the goto will branch to it so that when the
+   // code is split any fall-through will go to the right place.
+   // If there is no following block the goto will branch to the first block; in
+   // this case the goto should never be reached, it is there only to
+   // make sure that the instruction following the last real treetop will be in
+   // method's code, so if it is a helper call (e.g. for a throw) the return address
+   // is in this method's code.
+   //
+   TR::Compilation *comp = self()->comp();
+   TR::TreeTop * tt;
+   TR::Node * node;
+
+   if (lastBlock->getNumberOfRealTreeTops() == 0)
+      tt = lastBlock->getEntry();
+   else
+      tt = lastBlock->getLastRealTreeTop();
+
+   node = tt->getNode();
+
+   if (!(node->getOpCode().isGoto() ||
+         node->getOpCode().isJumpWithMultipleTargets() ||
+         node->getOpCode().isReturn()))
+      {
+
+      if (comp->getOption(TR_TraceCG))
+         {
+         traceMsg(comp, "%s Inserting goto at the end of block_%d\n", SPLIT_WARM_COLD_STRING, lastBlock->getNumber());
+         }
+
+      // Find the block to be branched to
+      //
+      TR::TreeTop * targetTreeTop = lastBlock->getExit()->getNextTreeTop();
+
+      if (targetTreeTop)
+         // Branch to following block. Make sure it is not marked as an
+         // extension block so that it will get a label generated.
+         //
+         targetTreeTop->getNode()->getBlock()->setIsExtensionOfPreviousBlock(false);
+      else
+         // Branch to the first block. This will not be marked as an extension
+         // block.
+         //
+         targetTreeTop = comp->getStartBlock()->getEntry();
+
+      // Generate the goto and insert it into the end of the last warm block.
+      //
+      TR::TreeTop *gotoTreeTop = TR::TreeTop::create(comp, TR::Node::create(node, TR::Goto, 0, targetTreeTop));
+
+      // Move reg deps from BBEnd to goto
+      //
+      TR::Node *bbEnd = lastBlock->getExit()->getNode();
+
+      if (bbEnd->getNumChildren() > 0)
+         {
+         TR::Node *glRegDeps = bbEnd->getChild(0);
+
+         gotoTreeTop->getNode()->setNumChildren(1);
+         gotoTreeTop->getNode()->setChild(0, glRegDeps);
+
+         bbEnd->setChild(0,NULL);
+         bbEnd->setNumChildren(0);
+         }
+
+      tt->insertAfter(gotoTreeTop);
+      }
+   }
+
 void OMR::CodeGenerator::findLastWarmBlock()
```

Comment:
I think the name of this function doesn't (or no longer) reflects what it is actually doing.  And it doesn't return anything as most of our "find" functions do.

Can you rename it to something like `prepareLastWarmBlockForCodeSplitting` ?

Reply:
> I think the name of this function doesn't (or no longer) reflects what it is actually doing

This function has not changed in any significant way in this PR. It's main effect is still to do this:

`lastWarmBlock->setIsLastWarmBlock(); `

I just took some code out and put into a separate function. The purpose of the remaining code is mostly to do the above (to mark the last warm block).

Reply:
I think it's important to reflect that before we call this function we don't know which block will be the splitting point. This function is looking for it.

Reply:
My point is that someone coming across this API will assume this function will find and return the last warm block in a method.  There is no API documentation (via Doxygen comments) provided for this function, so a developer will rely mostly on the name alone to make a determination on what it does.  "Inserting a goto" is not an operation one would expect if one was simply looking for the last block, as is implied by this function name.

My request is simply to rename it to something that more closely reflects what it does.  If you have an alternate suggestion for how the name could be improved or another means for providing clarity, please propose.

To be fair, we (which probably means "me") should have brought this up when the code was initially contributed many months ago, but didn't.  In the spirit of clarifying the code moving forward I am requesting this change now.

Reply:
We could just document the method. It's a bit hard to describe everything it does in the name. But I can rename as suggested.